src/main/scala/io/iohk/ethereum/blockchain/data/GenesisDataLoader.scala

```diff
@@ -128,21 +129,40 @@ class GenesisDataLoader(blockchain: Blockchain, blockchainConfig: BlockchainConf
     genesisData.alloc.zipWithIndex.foldLeft(initalRootHash) { case (rootHash, ((address, genesisAccount), _)) =>
       val mpt = MerklePatriciaTrie[Array[Byte], Account](rootHash, storage)
       val paddedAddress = address.reverse.padTo(addressLength, "0").reverse.mkString
+
       val stateRoot = mpt
         .put(
           crypto.kec256(Hex.decode(paddedAddress)),
           Account(
             nonce = genesisAccount.nonce
               .getOrElse(blockchainConfig.accountStartNonce),
             balance = genesisAccount.balance,
-            codeHash = genesisAccount.code.map(codeValue => crypto.kec256(codeValue)).getOrElse(Account.EmptyCodeHash)
+            codeHash = genesisAccount.code.map(codeValue => crypto.kec256(codeValue)).getOrElse(Account.EmptyCodeHash),
+            storageRoot = genesisAccount.storage.map(computeStorageRootHash).getOrElse(Account.EmptyStorageRootHash)
           )
         )
         .getRootHash
       stateRoot
     }
   }
 
+  private def computeStorageRootHash(storage: Map[UInt256, UInt256]): ByteString = {
```

Comment:
From what I saw. Usually the hash is computed when actually storing the key values. For instance after running a transaction which modified the data we can extract the resulting root hash. The problem here is that there is no way to know the root hash unless we actually build it. So I create a MPT in memory in order to get the hash. I did not find any existing helper to do this.

I see two alternative to do actually get the root hash : 

 - injecting the root hash (new argument in `loadGenesisData` with precomputed hashed). We could do this by first storing the data and getting the resulting root hash.
 - building the hash inside the function

I choose the second because this is similar to what we already do with the contract code (we compute the hash of the code, but it is indeed less involved for the code because there is no MPT).

At the same time it seems weird that the Genesis data loader stores hashes but not the code and the contract storage, so a small change the responsibilityies might be in order 🤔.

Reply:
So many of the details go over my head but it sounds like there's room for improvement. Do you reckon it's worth creating a refactoring ticket?

(IMHO we could also improve some of this stuff with decent factory methods, eg. `MerklePatriciaTrie.empty`.)